public/app/plugins/datasource/mixed/MixedDataSource.ts

```diff
@@ -39,23 +42,75 @@ export class MixedDatasource extends DataSourceApi<DataQuery> {
 
     // Build groups of queries to run in parallel
     const sets: { [key: string]: DataQuery[] } = groupBy(queries, 'datasource.uid');
-    const mixed: BatchedQueries[] = [];
+    const batches: BatchedQueries[] = [];
 
     for (const key in sets) {
-      const targets = sets[key];
-
-      mixed.push({
-        datasource: getDataSourceSrv().get(targets[0].datasource, request.scopedVars),
-        targets,
-      });
+      batches.push(...this.getBatchesForQueries(sets[key], request));
     }
 
     // Missing UIDs?
-    if (!mixed.length) {
+    if (!batches.length) {
       return of({ data: [] }); // nothing
     }
 
-    return this.batchQueries(mixed, request);
+    return this.batchQueries(batches, request);
+  }
+
+  /**
+   * Almost always returns a single batch for each set of queries.
+   * Unless the query is using a multi value variable.
+   */
+  private getBatchesForQueries(queries: DataQuery[], request: DataQueryRequest<DataQuery>) {
+    const dsRef = queries[0].datasource;
+    const batches: BatchedQueries[] = [];
+
+    // Using the templateSrv.replace function here with a custom formatter as that is the cleanest way
+    // to access the raw value or value array of a variable.
+    const datasourceUid = getTemplateSrv().replace(
+      dsRef?.uid,
+      request.scopedVars,
+      (value: string | string[], variable: CustomFormatterVariable) => {
+        // If it's not a data source variable, or single value
+        if (!Array.isArray(value)) {
+          batches.push({
+            datasource: getDataSourceSrv().get(queries[0].datasource, request.scopedVars),
+            queries,
+            scopedVars: request.scopedVars,
+          });
```

Comment:
I think we need to return here, or the for loop will loop over the value that could be string. Creating queries for every character in the UID. This can be observed if we're using a single value datasource variable.

Reply:
@oscarkilhed good catch!